### PR TITLE
enh: add stateless flag to numpy_function

### LINKED ISSUE #47777: enh: allow stateless numpy function
**System information**
- TensorFlow version (you are using): 2.4
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
`tf.numpy_function` is assumed to be stateful. However, many pure Python functions (as the example with `np.*` in the docs) are actually stateless. `numpy_function` calls `py_func_common` that actually has a 'stateful' argument. Can't we just pass this through?

**Will this change the current api? How?**
Add an argument `stateful` to `numpy_function`. Defaults to `True`, keeps backwards compatibility

**Who will benefit with this feature?**

This allows optimizations in the graph

**Any Other info.**
Is there any reason why this is not passed through currently?

### PR DESCRIPTION
**Pull request by jonas-eschle**
fixes #47777 

`numpy_function` can now be stateless.

Added test to ensure that the op is stateless/stateful depending on the flag.

### CODE REVIEW

**Comment by mihaimaruseac** [comment]
Since this changes public API, waiting for review from API owners first.

**Comment by ematejska** [comment]
The API owners are moving to a rotation based system instead of the meetings. Adding couple API Owners as reviewers.

File: tensorflow/python/ops/script_ops_test.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops_test.py:58]: Use self.assert instead of assert statements.

Comment (mdanatg) [inline tensorflow/python/ops/script_ops_test.py:58]: I don't think this is a robust operation. If you prefer, set up a test with a numpy function that's called twice, and check that the python function only tuns once - that's what the contract promises.

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops.py:704]: Please add a comment specifying that if you pass a stateful function with stateless=True, the behavior is undefined.

File: tensorflow/python/ops/script_ops_test.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops_test.py:66]: No need to specify int32 - it's the default.

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [reply tensorflow/python/ops/script_ops.py:704]: In particular, let's caution about mutating the input arguments.

File: tensorflow/python/ops/script_ops_test.py
Comment (mayou36) [reply tensorflow/python/ops/script_ops_test.py]: Indeed, I changed that to two calls

**Comment by mayou36** [comment]
I've changed everything as noted, thanks for the remarks

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops.py:704]: Typo: "may hav side effects"

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops.py:709]: Let's make a strnger distinction between "stateless operation" = thing your code might do that mutates some state and "stateful=False" = the user set this flac to False.

Specifically, the last sentence should say "Optimizations [...] are only performed when the `stateful` flag is set to `False`"

File: tensorflow/python/ops/script_ops_test.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops_test.py:46]: This should be `nonlocal`, not `global`.

File: tensorflow/python/ops/script_ops_test.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops_test.py:58]: No need for this override, I think (same below).

**Review by mdanatg — CHANGES_REQUESTED** [review]
Looks good. Just a couple more nits.

File: tensorflow/python/ops/script_ops.py
Comment (mayou36) [reply tensorflow/python/ops/script_ops.py:709]: Okay sure, I'll gladly change this. But the docs is taken one-by-one from `py_func_common`. Should this here be more "end user facing"? Or why would we not want to have the same docstring?

File: tensorflow/python/ops/script_ops_test.py
Comment (mayou36) [reply tensorflow/python/ops/script_ops_test.py:58]: Hm, true, if the nodes generated by the `nupmy_function` with the same Python function are considered the same

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [reply tensorflow/python/ops/script_ops.py:709]: Definitely. Unlike py_func_common, this docstring will appear on the tensorflow.org website in the public documentation.

File: tensorflow/python/ops/script_ops_test.py
Comment (mdanatg) [reply tensorflow/python/ops/script_ops_test.py:50]: Sounds good, that's a func change as well.

But I was referring to the `autograph=False` override. The test should run with just `@def_function.function` as well.

File: tensorflow/python/ops/script_ops.py
Comment (mayou36) [reply tensorflow/python/ops/script_ops.py]: I see, sure. I rewrote the docs completely to make it clearer, let me know what you think

File: tensorflow/python/ops/script_ops_test.py
Comment (mayou36) [reply tensorflow/python/ops/script_ops_test.py]: :) good, done that as well

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops.py:705]: Suggest: "Setting this argument to False tells the runtime to treat the function as stateless, which enables certain optimizations."

**Review by mdanatg — APPROVED** [review]

**Comment by mdanatg** [comment]
Thanks!

File: tensorflow/python/ops/script_ops.py
Comment (mayou36) [reply tensorflow/python/ops/script_ops.py:705]: Perfect, changed

**Comment by mayou36** [comment]
Thanks for the review and all the work with the library!

**Review by mdanatg — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by mdanatg** [comment]
Looks like a couple of tests are failing - check the logs for Ubuntu CPU for example.

**Comment by mayou36** [comment]
So one if from the API compatibility test: since we introduce a new option in the function, this indeed changes the API. But I am not more familiar with why this fails.

For the other one, it basically tells that the statelessness does not work...

Can you rerun the tests?

**Comment by mdanatg** [comment]
For the API compatibility test, the test logs should give you the commands to run to update the tests.

**Comment by mayou36** [comment]
> For the API compatibility test, the test logs should give you the commands to run to update the tests.

It doesn't, as far as I can see (https://source.cloud.google.com/results/invocations/eb308cd4-fcad-4909-8e89-4de24d817093/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/tests;group=__main__.ApiCompatibilityTest;test=testAPIBackwardsCompatibility;row=1)

It just says that it failed

**Comment by mdanatg** [comment]
Sorry, those details are in the test log, I should have clarified it: https://source.cloud.google.com/results/invocations/eb308cd4-fcad-4909-8e89-4de24d817093/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/log (check the tabs at the top: Test | Target Log | Target Details | Artifacts).

Here's the relevant piece:

```
ERROR:tensorflow:TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.
```

**Comment by mayou36** [comment]
Ah indeed, sorry, yet unfamiliar with the TF CI! Thanks, I'll do that

**Comment by gbaned** [comment]
@mayou36  Any update on this PR? Please. Thanks!

**Comment by mayou36** [comment]
Indeed, I've updated the golden rule. Can you run the workflow again?

**Review by mdanatg — APPROVED** [review]

**Comment by aselle** [comment]
this makes numpy_function the same as py_func in v1 was. This change was explicit, because it is almost always a bad idea to do stateful things. I don't see how this is an enhancement right now. Can you explain what your goals are with this change in terms of use-cases?

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [inline tensorflow/python/ops/script_ops.py:644]: Sorry, missed this bit - please make sure the flag defaults to False. Otherwise it breaks backward compatibility.

**Review by mdanatg — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/script_ops.py
Comment (mdanatg) [reply tensorflow/python/ops/script_ops.py:644]: Nevermind, I see the old code defaulted to True as well.

**Comment by jonas-eschle** [comment]
> this makes numpy_function the same as py_func in v1 was. This change was explicit, because it is almost always a bad idea to do stateful things. I don't see how this is an enhancement right now. Can you explain what your goals are with this change in terms of use-cases?

Exactly. This is the goal. `numpy_function` _is_ currently stateful, the goal is to have a flag to allow to make it stateless.

There are two functions, `py_function` and `numpy_function`. The former keeps track of the gradients and stuff that happens inside (AFAIU it's an "eager wrapper") and can accumulate memory (as explained here: https://github.com/tensorflow/tensorflow/issues/35010#issuecomment-798198842), while the latter does not (AFAIU it's a "pure Python, no TF"). Unfortunately, `numpy_function is always considered stateful, although the internal function actually takes a "stateful" flag. Instead of making this always to be true, the proposed change allows to pass this flag through to enable a stateless `numpy_function`.

The current problem on this PR is: how to _test_ this. Actually, I would suggest to not test the actual function. Because it is just about passing through a flag and giving the runtime a guarantee. Testing that guarantee would need internal knowledge.

**Review by aselle — APPROVED** [review]

**Review by mdanatg — APPROVED** [review]

**Comment by mdanatg** [comment]
Sorry, there's a change that we missed and was caught by the internal checks - could you add a short bullet to https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md, under "Bug Fixes and Other Changes", "TF Core"? And on this occasion, you might wish to add your name under the Contributors list as well.

**Comment by gbaned** [comment]
@jonas-eschle Can you please check @mdanatg's comments and keep us posted ? Thanks!

**Comment by jonas-eschle** [comment]
Thanks, I've done that (not pushed yet), and finally fixed the tests.

One thing:
> you might wish to add your name under the Contributors list as well.

Sure, but where exactly? I've tried to find a place (a "CONTRIBUTORS" file?) but couldn't find anything.

**Comment by jonas-eschle** [comment]
Okay, can you please run the CI again? It is ready from my side.
I've added my name, but there was no other name. Is it correct like that? Otherwise, feel free to modify directly

**Comment by gbaned** [comment]
@jonas-eschle Can you please resolve conflicts? Thanks!

**Comment by jonas-eschle** [comment]
Done, ready from my side

**Review by mdanatg — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by jonas-eschle** [comment]
@gbaned resolved them and moved to 2.8 release

**Review by mdanatg — APPROVED** [review]

**Comment by gbaned** [comment]
@jonas-eschle Can you please resolve conflicts? Thanks!

**Comment by jonas-eschle** [comment]
Resolved the conflicts and moved it into the 2.8 changelog

**Review by mdanatg — APPROVED** [review]

**Pull request merged**
